app/src/main/java/com/alphawallet/app/viewmodel/DappBrowserViewModel.java

```diff
@@ -350,26 +351,63 @@ public String getSessionId(String url)
         return Uri.parse(uriString).getUserInfo();
     }
 
-    public void addCustomChain(WalletAddEthereumChainObject chainObject) {
-        this.ethereumNetworkRepository.saveCustomRPCNetwork(chainObject.chainName, extractRpc(chainObject), chainObject.getChainId(),
-                chainObject.nativeCurrency.symbol, "", "", false, -1L);
+    public boolean addCustomChain(WalletAddEthereumChainObject chainObject) {
+        String rpc = extractRpc(chainObject);
+        String blockExplorer = extractBlockExplorer(chainObject);
+        if (rpc == null) return false;
+
+        this.ethereumNetworkRepository.saveCustomRPCNetwork(chainObject.chainName, rpc, chainObject.getChainId(),
+                chainObject.nativeCurrency.symbol, blockExplorer, "", false, -1L);
 
         tokensService.createBaseToken(chainObject.getChainId())
                 .subscribeOn(Schedulers.io())
                 .observeOn(Schedulers.io())
                 .subscribe(w -> { }, e -> { })
                 .isDisposed();
+
+        return true;
+    }
+
+    private String extractBlockExplorer(WalletAddEthereumChainObject chainObject)
+    {
+        for (String thisRpc : chainObject.blockExplorerUrls)
+        {
+            if (isValidUrl(thisRpc)) //ensure RPC doesn't contain malicious code
```

Comment:
Suggest adding unit test for `isValidUrl` with the malicious payload to ensure it won't pass the gate.